Niraj:Narcissistic number computation, test coverage implementation

nirajunify opened the pull request:
This commit/PR consists of:
- Narcissistic numbers computation and test
- Added Code coverage configuration in jest.config.js
- Added coverage directory in .gitignored
- Added source map support in .tsconfig.json
- Added Jest Config article link in references
- Removed dummy foo.ts and its test file

nirajunify committed:
Narcissistic number computation, test coverage implementation
This commit consists of:
- Narcissistic numbers computation and test
- Added Code coverage configuration in jest.config.js
- Added coverage directory in .gitignored
- Added source map support in .tsconfig.json
- Added Jest Config article link in references
- Removed dummy foo.ts and its test file

nirajunify committed:
Added spacing (for better formatting) in narcissistic number computation function

krajasekhar-unify commented on src/niraj/narcissisticNumbers.spec.ts:
```diff
@@ -0,0 +1,60 @@
+});
+
+describe('When no range given', () => {
+  test('it should return an empty array', () => {
```
Isn't the purpose of `describe` is to group the tests and have some common scope and it's own `beforeEach` and `afterEach` than using it for each tests. Having `describe` for each test won't give any advantage right?

krajasekhar-unify replied:
https://jestjs.io/docs/en/setup-teardown#scoping

nirajunify replied:
Yes. What you said is absolutely correct. In this instance, I chose to use describe and test to break a long test statement.

nirajunify replied:
I could have written the tests without 'describe', but adding 'describe' helped my test suits look good

krajasekhar-unify replied:
Since all the tests are related to testing a single functionality, can write a single describe block to enclose all the tests.

nirajunify replied:
> Since all the tests are related to testing a single functionality, can write a single describe block to enclose all the tests.

Perhaps, Can you please give recommendation on how to arrange my tests in logical grouping?

krajasekhar-unify replied:
how about something like this?
```
describe('Narcissistic number', () => {
   test('0 is not a narcissistic number', () => {... });
   test('it should return an empty array for a negative starting number input', () => {... });
   ...
});
```

ramjayunify replied:
describe can be used to group test cases of similar kinds. In this case @nirajunify has created a detailed test cases, each one representing different types or edge cases. May be we can group those things into a nested describe such as negative numbers in one group
```
describe('Given a negative number', () => {
  describe('Given a negative starting number', () => {
    test('it should return an empty array', () => {
      expect(narcissisticNumbers(-10).length).toBe(0);
    });
  });

  describe('Given a negative ending number', () => {
    test('it should return an empty array', () => {
      expect(narcissisticNumbers(0, -10).length).toBe(0);
    });
  });
});
```

Or when we are testing for different range of positive numbers like this

```
describe('When starting number is 10 and ending number is 10000', () => {
  const narcissisticNumbersList = [153, 370, 371, 407, 1634, 8208, 9474];
  test('it should return an array of numbers [153, 370, 371, 407, 1634, 8208, 9474]', () => {
    expect(narcissisticNumbers(10, 10000).length).toEqual(
      narcissisticNumbersList.length
    );
    expect(narcissisticNumbers(10, 10000)).toEqual(narcissisticNumbersList);
  });

  test('it should return an array of numbers [153, 370, 371, 407]', () => {
    expect(narcissisticNumbers(10, 500).length).toEqual(
      narcissisticNumbersList.splice(0,4).length
    );
    expect(narcissisticNumbers(10, 500)).toEqual(narcissisticNumbersList.splice(0,4));
  });
});
```
But I read a blog, where it says grouping test cases using describe is not the best way to do it.
Please refer this [https://kentcdodds.com/blog/avoid-nesting-when-youre-testing](url)

nirajunify replied:
Thank you @krajasekhar-unify , @ramjayunify  for your valuable feedback. I will update the tests

zulqer-nain replied:
I think in each describe group there must be at least one pass case and one fail case .

krajasekhar-unify replied:
Thanks @ramjayunify for the `kentcdodds` article. I am convinced to not group the tests with describe at all. Do we have anything to inspire from the article and define in our guidelines?

nirajunify replied:
Thanks. I will update the specs

praveen-unify commented on src/niraj/narcissisticNumbers.ts:
```diff
@@ -0,0 +1,40 @@
+ */
+
+export const narcissisticNumbers = (start?: number, end?: number): number[] => {
+  const result = [];
```
we can define the type for 'result' as well.

nirajunify replied:
Agreed. I will update the code

praveen-unify commented on src/niraj/narcissisticNumbers.ts:
```diff
@@ -0,0 +1,40 @@
+
+  if (!start || !end || start < 0 || end < 0 || end < start) return result;
+
+  for (let i = start; i <= end; i++) {
```
for 'i' also we can define it's type.

nirajunify replied:
No necessarily. Since I am defining the variable and immediately assigning the value, typescript can automatically infer that variable's type - I don't need to define the type.

praveen-unify commented on src/niraj/narcissisticNumbers.ts:
```diff
@@ -0,0 +1,40 @@
+  // All numbers between 1-10 are narcissistic number
+  if (numberToCheck > 0 && numberToCheck < 10) return true;
+
+  const strNum = String(numberToCheck);
```
We can define the type of 'strNum'.

krajasekhar-unify replied:
Please help to clarify my doubt. Do we need to do type check for every intermediate value?
- If we are using any js functions which always outputs a particular type (String -> string, Math.sqrt -> number), it already knows the type right?
- For the below code, even though type is not declared for `varB` and `varA`, at the time of using it editor is still showing the type as number as it reflected from the return type of the function? Do we still need to define types for immediate variables?
```
const myFunction = (): number => 4;
const varB = myFunction();
const varA = varB + 3;
console.log(varA, varB);
```

nirajunify replied:
You are right Raj. Wherever typescript can automatically infer a variable's type, you do not have to provide its type.

praveen-unify commented on src/niraj/narcissisticNumbers.ts:
```diff
@@ -0,0 +1,40 @@
+  const strNum = String(numberToCheck);
+
+  return (
+    numberToCheck ==
```
Please use 'strict equality' using === instead of ==

praveen-unify reviewed (commented):
I have some feedback to changes.

nirajunify committed:
Test cases updated, prettier config updated for windows

Based on the code reviews, I updated the code to
- apply types to variables wherever required
- replace == with ===
- updated prettier config to resolve windows specific file watching issue
- updated the test cases for better logical grouping

nirajunify committed:
Resolved conflict with master branch, reason issue5

praveen-unify commented on .gitignore:
```diff
@@ -103,3 +103,8 @@ dist
+src/**/*.map
+coverage
+=======
+>>>>>>> 5d7f7554696eab1be0365ee5a4ca666322d36243
```
Conflict is not resolved, please resolve it.

nirajunify replied:
It's been resolved now

nirajunify committed:
Narcissistic number checker and its tests updated
Added negative testing for narcissistic number
removed unnecessary checking for number 0

nirajunify committed:
Refactored the code the remove duplicate test cases and to add 0 as narcissistic number

nirajunify committed:
Removed a duplicate test case
